### PR TITLE
Drop user agent matches and a bugfix for over-committing KafkaRecords too early

### PR DESCRIPTION
**Pull request by gregschohn**
### Description

Add the ability to drop captures, while still forwarding the contents, for requests when a specific header matches some regex.

Since captures are really on a connection (socket) basis and not a request one, there are additional complications.
Data is captured as it is received.  We may not know that we want to drop packets early enough.  To compensate and to give the traffic stream enough details to differentiate this occurrence from a bug where packets were dropped, a new "RequestIntentionallyDropped" TrafficObservation has been introduced.  This observation will only be used when data for a request has been included.  That means that if we get a packet that has the entire HTTP preamble (first line + headers) that matches the drop filter, we won't capture anything at all.
The --suppressCaptureForHeaderMatch argument has been added to the CaptureProxy.  It expects two elements for a header name and a regex to match that header's value.  When a request matches, it will be passed to the HTTP server, but will NOT be captured as per the above.  Because the capture proxy only captures a handful of headers so that it can understand request boundaries, the predicate to scan additional headers must tweak the set of headers that the LoggingHttpRequestHandler's internal classes will preserve.

There was also a bugfix for over-committing KafkaRecords too early and a spurious log message has been disabled.

* Category New feature, Bug fix, Test fix
* Why these changes are required? To stop the capture proxy from capturing data that is being pulled by the fetch migration.
* What is the old behavior before changes and new behavior after changes?  Now users can suppress some captures with additional command line flags.  Otherwise, behavior is identical.

### Issues Resolved
https://opensearch.atlassian.net/browse/MIGRATIONS-1444

### Testing
`./gradlew :dockerSolution:composeUp` ... OSB --test script and doc count check

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: TrafficCapture/trafficCaptureProxyServer/src/main/java/org/opensearch/migrations/trafficcapture/proxyserver/CaptureProxy.java [hunk 1]
```diff
@@ -274,6 +286,14 @@ private static SslContext loadBacksideSslContext(URI serverUri, boolean allowIns
         }
     }
 
+    private static Map<String, String> convertPairListToMap(List<String> list) {
+        var map = new TreeMap<String, String>();
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficCaptureProxyServer/src/main/java/org/opensearch/migrations/trafficcapture/proxyserver/CaptureProxy.java:290]: Beust takes care of that for us already & the user will see the usage.  Here's what the user will see with the following command args...
`--kafkaConnection kafka:9092 --otelCollectorEndpoint http://localhost:4317 --destinationUri https://localhost:19200 --insecureDestination --listenPort 59200 --suppressCaptureForHeaderMatch user-agent foo --suppressCaptureForHeaderMatch cookie`

```
Starting Capture Proxy
Got args: --kafkaConnection; kafka:9092; --otelCollectorEndpoint; http://localhost:4317; --destinationUri; https://localhost:19200; --insecureDestination; --listenPort; 59200; --suppressCaptureForHeaderMatch; user-agent; foo; --suppressCaptureForHeaderMatch; cookie
Expected 2 values after --suppressCaptureForHeaderMatch
Got args: --kafkaConnection; kafka:9092; --otelCollectorEndpoint; http://localhost:4317; --destinationUri; https://localhost:19200; --insecureDestination; --listenPort; 59200; --suppressCaptureForHeaderMatch; user-agent; foo; --suppressCaptureForHeaderMatch; cookie
Usage: <main class> [options]
  Options:
    --destinationConnectionPoolSize
...
```

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/RequestResponsePacketPair.java [hunk 2]
```diff
@@ -23,17 +23,17 @@ public enum ReconstructionStatus {
 
     HttpMessageAndTimestamp requestData;
     HttpMessageAndTimestamp responseData;
+    final ITrafficStreamKey trafficStreamKey;
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/RequestResponsePacketPair.java:26]: What do you mean by original?  Either way, I don't really have a good answer from looking at the code to know what it's for.  I've renamed this to 'firstTrafficStreamKeyForRequest'.  Does that make it more clear?

File: TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java [hunk 3]
```diff
@@ -204,6 +205,16 @@ public CompletableFuture<T> flushCommitAndResetStream(boolean isFinal) throws IO
         return future;
     }
 
+    @Override
+    public void cancelCaptureForCurrentRequest(Instant timestamp) throws IOException {
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java:209]: That is correct.  If a caller wants to make sure that there's no trace, it's reasonable that they'll be able to send in a request with the filtering logic early enough (first packet).  That said, we don't parse the headers as they come in, but rather once they've all arrived.  Doing them as they come in would be a really nice to have.

However, there may be cases that are spread across time.  We have two choices.  Buffer and manage, or offload ASAP (keeping in mind that there's a lot of buffering through the stacks, just not super easy to retract).  Not capturing is an optimization on the wire protocol, but at the expense of compute & memory for the proxy.  Not replaying is the visible high-level requirement, which we'll meet by adding the tombstone.  Trying to do more for what could be very rare cases, considering that we're going to be setup to log nearly all of the traffic anyway doesn't seem like an investment that would ever pay off.  (as much as it pains me to NOT do that wire optimization)

File: TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java [hunk 4]
```diff
@@ -204,6 +205,16 @@ public CompletableFuture<T> flushCommitAndResetStream(boolean isFinal) throws IO
...
+        beginSubstreamObservation(timestamp, TrafficObservation.REQUESTDROPPED_FIELD_NUMBER, 1);
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java:210]: Good question - as of when you reviewed this, No.  I've pushed a new commit.
I had gotten distracted yesterday when I had published it & had forgotten to walk through the unit tests to make sure that they were complete (they weren't & the broke the same way that the code broke).

File: TrafficCapture/nettyWireLogging/src/test/java/org/opensearch/migrations/trafficcapture/netty/ConditionallyReliableLoggingHttpRequestHandlerTest.java [hunk 5]
```diff
@@ -109,22 +118,116 @@ private static byte[] consumeIntoArray(ByteBuf m) {
...
+        return w -> {
+            for (int i = 0; i< fullTrafficBytes.length; ++i) {
                 var singleByte = TestUtilities.getByteBuf(Arrays.copyOfRange(fullTrafficBytes, i, i+1), usePool);
                 w.writeInbound(singleByte);
             }
-        });
+        };
+    }
+
+    // This test doesn't work yet, but this is an optimization.  Getting connections with only a
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/nettyWireLogging/src/test/java/org/opensearch/migrations/trafficcapture/netty/ConditionallyReliableLoggingHttpRequestHandlerTest.java:141]: Yes.

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/kafka/KafkaProtobufConsumer.java
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/kafka/KafkaProtobufConsumer.java:1]: It wasn't deleted when the branch point had happened for another brach that I might have merged into this one.  Either way, it had come back in and I had to delete it again here.

**Review by lewijacn — APPROVED** [review]

**Pull request merged**